### ISSUE TITLE
How to block DNS requests locally

### ISSUE THREAD

**Issue by Tryfos** [rospogrigio/localtuya#407]
In the docs it is advised to also block DNS requests locally to the router, in order to totally block internet to tuya clients and not having them in a zombie state.
I searched through router firewall options, I can only block port 53 from LAN to WAN, not from LAN to LAN. Can someone explain what is the easiest way to do that?

**Comment by imkebe** [comment]
It might not be possible with common user network devices. You need at least be able to manage to block request from specific devices to your router port 53. However you can try to change network architecture.
1. Put DHCP server at your HA device.
2. Set HA to be static IP.
3. Configure it like it used to be on your router, however remember to distingush default gateway IP and DHCP server IP. You may want to set specific DHCP Options like (5,6 - your HA IP, 3 - you router IP).
4. Disable DHCP assigment  on your router.
5a. Implement iptables block. Example : ```/sudo iptables -A INPUT -i eth0 -m mac --mac-source MAC-ADDRESS-OF-TUYA-DEVICE -p udp --dport 53 -j DROP```
5b. Install adguard home and block devices using it's settings

**Comment by Tryfos** [comment]
I see. Thank you. I was hoping of a somewhat simpler solution...

**Comment by imkebe** [comment]
> I see. Thank you. I was hoping of a somewhat simpler solution...
> Also I don't understand in step 3: "You may want to set specific DHCP Options like (5,6 - your HA IP, 3 - you router IP)".

https://www.incognito.com/tutorials/dhcp-options-in-plain-english/

**Comment by Tryfos** [comment]
Didn't know about RFC-defined DHCP and still can't understand its value. Anyway, there are no such options in my router's GUI.
After you blocked DNS locally and Internet, do your plugs report power? As soon as I block Internet, reporting of power stops (switch works). I can't understand how it'll work after I block DNS, since it stops reporting immediately after I block internet.

**Comment by imkebe** [comment]
> Didn't know about RFC-defined DHCP and still can't understand its value. Anyway, there are no such options in my router's GUI.

You won't be able to do so using CPE grade router. That's why you need to disable your router dhcp service and install DHCP server (isc-dhcp-server, bind, etc.) outside - it might be a HA device.

> After you blocked DNS locally and Internet, do your plugs report power? As soon as I block Internet, reporting of power stops (switch works). I can't understand how it'll work after I block DNS, since it stops reporting immediately after I block internet.
I don't have power plugs. However reporting and connection problems are not the same. I just ensure zero random remote network policy for my iot devices.
What i can see internally it's that tuya devices in my location try to resolve some tuya.eu based domains.

**Comment by postlund** [comment]
> Didn't know about RFC-defined DHCP and still can't understand its value. Anyway, there are no such options in my router's GUI.
> After you blocked DNS locally and Internet, do your plugs report power? As soon as I block Internet, reporting of power stops (switch works). I can't understand how it'll work after I block DNS, since it stops reporting immediately after I block internet.

A lot of people are seeing that power reporting stops working after blocking internet access. I'm working on a polling version to fix that.

**Comment by Tryfos** [comment]
> > Didn't know about RFC-defined DHCP and still can't understand its value. Anyway, there are no such options in my router's GUI.
> > After you blocked DNS locally and Internet, do your plugs report power? As soon as I block Internet, reporting of power stops (switch works). I can't understand how it'll work after I block DNS, since it stops reporting immediately after I block internet.
> 
> A lot of people are seeing that power reporting stops working after blocking internet access. I'm working on a polling version to fix that.

Thanks for the great work, this is a great integration btw! Bear in mind that in a while this is going to be our only "hope" to disengage from the cloud, since the trend for the suppliers is to completely avoid ESP chips (only 2/10 plugs I  recently bought had flashable ESP and I had done my research). 
I thought that the "solution" to power readings problem after internet blocking would be to also block local DNS requests from the plugs to router per official docs. But it seems a little odd to me that such a solution would fix this, since power readings are **_immediately_** stopped when internet is blocked. I don't see how blocking the DNS requests after that would fix it.
Btw the interesting thing is that the switching part (both on/off and status report) **work ok** for me even after blocking internet. Plugs never go "offline" or "unavailable". Only power readings stop.

**Comment by postlund** [comment]
Yeah, until there's a solution to flash some other software to the new SoC:s used by most Tuya devices nowadays (I think it's one from Realtek), we will be stuck to the local interface or the cloud. We might reach to that point at some time, but it's probably a long time from now.

From what I have gathered, the tuya platform allows developers to push updates to the local interface, local and cloud *or* cloud-only. So it's likely the case that energy statistics are cloud-only. That makes sense in many cases as total consumption (I.e.kWh) are usually what's most interesting and that calculation is done by the cloud. Most plugs only publish voltage, current and current consumption.

**Comment by mycanaletto** [comment]
I use a Windows DNS (Active Directory), it is possible to assign a particular DNS server for each reservation, and it is enough to put it 127.0.0.1 so that it is not seen by Tuya. I suppose this is possible on any slightly advanced DHCP server under Linux.

![image](https://user-images.githubusercontent.com/45459269/114302843-3600e900-9acb-11eb-93cc-1307dde0d767.png)

**Comment by OdinGitDat** [comment]
>     5. Implement iptables block. Example : `/sudo iptables -A INPUT -i eth0 -m mac --mac-source MAC-ADDRESS-OF-TUYA-DEVICE -p udp --dport 53 -j DROP`
@imkebe Would this command work as a dd-wrt firewall command?
